### DISCUSSION TITLE
How to terminate a state machine that a user is executing?

### DISCUSSION THREAD

**Discussion by LamGC** [rubenlagus/TelegramBots#933] [Q&A] [answered]
I try to add a "Cancel" command to terminate the ReplyFlow that has been executed. But I don't know how to stop the ReplyFlow that a user is running. What should I do to stop the ReplyFlow that a user is running?
(For example, delete some information in dB, or perform an operation.)

**Comment by LamGC** [comment]
Guys, I find that if I clear the list provided by `bot.replies()` in the action of reply, I can stop all the following processes, but is this appropriate ? ( I mean, what's the bad effect on AbilityBot )

**Reply by addo47** [reply]
That would not allow the bot to trigger any reply logic to any user. That is iterated on for when the bot processes new messages!

**Comment by addo47** [comment, marked as answer]
`db.<Long, Integer>getMap("user_state_replies").remove(<CHAT_ID>)` should reset the state. I agree that a native solution for this should be present. Can you open an issue to track this?

**Reply by LamGC** [reply]
I tried to clear this key in one step of replyflow, but it didn't seem to work. Replyflow can still run. :O

**Reply by addo47** [reply]
Inside a flow, it would definitely not work. It has to be done outside the flow. Can you explain your flow here? You could technically create a terminal reply here if the user says anything that breaks the logic.

**Reply by LamGC** [reply]
> Inside a flow, it would definitely not work. It has to be done outside the flow. Can you explain your flow here? You could technically create a terminal reply here if the user says anything that breaks the logic.

I try to execute this code in another ability (`/cancel`) instead of running it in the replyflow process that is being run.

This is my test code: 
```java
    public Ability test() {
        ReplyFlow botHello = ReplyFlow.builder(db())
                .onlyIf(upd -> "/test hello".equalsIgnoreCase(upd.getMessage().getText()))
                .action((bot, update) -> bot.silent().send("Hello!", update.getMessage().getChatId()))
                .next(ReplyFlow.builder(db)
                        .action((bot, upd) -> {
                            bot.silent().send("who are you?", upd.getMessage().getChatId());
                        })
                        .onlyIf(update -> "hello".equalsIgnoreCase(update.getMessage().getText()))
                        .next(Reply.of((bot, upd) -> bot.silent()
                                .send("I see. You're " + upd.getMessage().getText().substring("I am ".length()), upd.getMessage().getChatId()),
                                upd -> upd.getMessage().getText().startsWith("I am ")))
                        .build())
                .build();

        return Ability.builder()
                .name("test")
                .info("A test.")
                .locality(Locality.ALL)
                .privacy(Privacy.PUBLIC)
                .enableStats()
                .action(ctx -> silent.send("In action.", ctx.chatId()))
                .reply(botHello)
                .reply(Reply.of((bot, upd) -> silent.send("Reply 01", upd.getMessage().getChatId())))
                //.reply(Reply.of((bot, upd) -> silent.send("Reply 02", upd.getMessage().getChatId()),
                //        update -> "/test 123".equals(update.getMessage().getText())))

                .build();
    }

    public Ability cancel() {
        return Ability.builder()
                .name("cancel")
                .info("Cancel current operation.")
                .privacy(Privacy.PUBLIC)
                .locality(Locality.ALL)
                .action(ctx -> {
                    db.<Long, Integer>getMap("user_state_replies").remove(ctx.chatId());
                    silent.send("Operation cancelled.", ctx.chatId());
                })
                .build();
    }
```

**Reply by addo47** [reply]
The cancel ability won't ever run  since you have a general reply defined above:
`.reply(Reply.of((bot, upd) -> silent.send("Reply 01", upd.getMessage().getChatId())))`

This line will get triggered on every message. AbilityBot doesn't process commands if any reply is executed (a design choice that I disagree with today, I think I might issue a PR for this soon).

So, remove that line and attempt to cancel. It should be fine.

**Reply by addo47** [reply]
The other suggestion is to terminate the flow. You can append another `.next` reply to the flow, in that flow you could have an empty action (or something that prints/sends a message) and the flags would be the counter condition. So if you're expecting a user to say "I am Name", then you could do `!msg.startsWith("I am")`. Then the reply will execute and it would be a terminal reply (since no more `next` replies are declared attached to it).

Let me know if this is unclear, I can provide an example.

**Reply by LamGC** [reply]
> The other suggestion is to terminate the flow. You can append another `.next` reply to the flow, in that flow you could have an empty action (or something that prints/sends a message) and the flags would be the counter condition. So if you're expecting a user to say "I am Name", then you could do `!msg.startsWith("I am")`. Then the reply will execute and it would be a terminal reply (since no more `next` replies are declared attached to it).
> 
> Let me know if this is unclear, I can provide an example.

I understand, that for a short replyflow, it's a great idea to cancel the current replyflow by using the "empty operation" reply, but it may not be applied to a longer process replyflow (it may be very complicated).

**Reply by LamGC** [reply]
>The cancel ability won't ever run since you have a general reply defined above:
>`.reply(Reply.of((bot, upd) -> silent.send("Reply 01", upd.getMessage().getChatId())))`
>
>This line will get triggered on every message. AbilityBot doesn't process commands if any reply is executed (a design choice that I disagree with today, I think I might issue a PR for this soon).
>
>So, remove that line and attempt to cancel. It should be fine.

It works, because the unconditional reply does not stop the ReplyFlow.

Thank you very much for your reply!
